### DISCUSSION TITLE
HTTP In Memory Response Cache

### DISCUSSION THREAD

**Discussion by michal-ciechan** [reactiveui/refit#1774] [Ideas]
We need an HTTP client that leverages ASP.NET Core Response and/or Output caching (which uses relatively standard HTTP caching headers).

Would this be something you would accept into Refit or should we build on top of/to the side of Refit?

We are initially looking to support the `If-Modified-Since` caching mechanism, in short:

- only affect certain requests (by default same ones as [ASP.NE Core OutputCaching Default output caching policy](https://learn.microsoft.com/en-us/aspnet/core/performance/caching/output?view=aspnetcore-8.0#default-output-caching-policy))
- For requests in question use an in-memory cache (probably [Microsoft.Extensions.Caching.Memory.MemoryCache](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.memory.memorycache))
- send a request with `If-Modified-Since` for requests in question where we have a cached entry
- If `304` is returned, return cached response, otherwise store the result adhering to cache headers, and return it.

**Comment by JeremyBP** [comment]
Hi,

Just to let you know that I just pushed Apizr v6.1.0 which includes your response caching feature on top of Refit.
You just have to pick the new SetByHeader cache mode to let the server control the client cache mode and lifetime, depending on response headers such as `Cache-Control`, `Expires`, `ETag`/`If-None-Match` and `Last-Modified`/`If-Modified-Since`.
More info on [Apizr documentation](https://www.apizr.net/articles/config_datacaching.html).

Let me know if it suits your needs or if I forgot something :)

**Comment by glennawatson** [comment]
HTTP caching is out of scope for Refit's core - Refit is a thin typed wrapper over HttpClient and intentionally stays out of the response pipeline beyond serialization. The right place for conditional-request / 304 / If-Modified-Since handling is a `DelegatingHandler`, which composes cleanly with any Refit client.

As JeremyBP noted, this is already available built on top of Refit: Apizr v6.1.0+ ships exactly this via its `SetByHeader` cache mode, honoring `Cache-Control`, `Expires`, `ETag`/`If-None-Match`, and `Last-Modified`/`If-Modified-Since`. See https://www.apizr.net/articles/config_datacaching.html

Recommendation: use Apizr, or write a small caching `DelegatingHandler` (backed by `IMemoryCache`) and plug it into your Refit client's handler chain. Closing as addressed by existing tooling.
